test/test_type_promotion.py

```diff
@@ -138,39 +139,50 @@ def test_mixed_type_backward(self):
         f_dtypes = [torch.float, torch.double]
         f_dtypes = f_dtypes if self.device == 'cpu' else f_dtypes + [torch.half]
         i_dtypes = [torch.int, torch.long]
-        for f in ['add', 'sub', 'rsub', 'mul', 'div']:
-            for dtype1 in f_dtypes:
-                for dtype2 in (f_dtypes + i_dtypes):
-                    x = torch.ones(10, requires_grad=True, dtype=dtype1, device=self.device)
-                    y = torch.ones(10, dtype=dtype2, device=self.device)
-
-                    func = getattr(torch, f)
-                    func(x, y).sum().backward()
-
-    # verifies that a.add(b) is the same as a.to(b.dtype).add(b) in cases
-    # where that should hold.
+        for func in [torch.add, torch.sub, torch.rsub, torch.mul, torch.div]:
+            for dtype1, dtype2 in itertools.product(f_dtypes, f_dtypes + i_dtypes):
+                x = torch.ones(10, requires_grad=True, dtype=dtype1, device=self.device)
+                y = torch.ones(10, dtype=dtype2, device=self.device)
+                func(x, y).sum().backward()
+
+    # verifies that torch.<op>(first, second) is the same as 
+    # torch.<op>(first.to(common_dtype), second.to(common_dtype)) in cases where that should hold.
     def test_many_promotions(self):
-        from_to = {
-            torch.float16: torch.float32,
-            torch.half: torch.float16,
-            torch.int: torch.long,
-            torch.uint8: torch.long,
-            torch.uint8: torch.float,
-            torch.int: torch.float,
-            torch.int: torch.double,
-            torch.int16: torch.long,
-            torch.float16: torch.double,
-            torch.bool: torch.long,
-            torch.bool: torch.float
-        }
-
-        for k, v in from_to.items():
-            a = torch.rand([3, 3], device=self.device).to(k)  # no _th_uniform for half on cpu.
-            b = torch.rand([3, 3], device=self.device).to(v)
-            c = a.add(b)
-            d = a.to(v).add(b)
-            self.assertEqual(c.dtype, d.dtype, message='from {} to {}'.format(k, v))
-            self.assertEqual(c, d, message='from {} to {}'.format(k, v))
+        # Can also include half on CPU in cases where it will be promoted to a
+        # supported dtype
+        dtypes1 = torch.testing.get_all_math_dtypes('cuda')
+        dtypes2 = torch.testing.get_all_math_dtypes(self.device)
+        ops = [torch.add, torch.sub, torch.mul, torch.div, torch.rsub]
+        for dt1, dt2 in itertools.product(dtypes1, dtypes2):
+            for op, non_contiguous in itertools.product(ops, [True, False]):
+                common_dtype = torch.promote_types(dt1, dt2)
+                if common_dtype == torch.half:
+                    continue
```

Comment:
Why is this? Is it because this means we didn't do any type promotion?

Reply:
right, if we're doing add(half_tensor, float_tensor) the common dtype is float and is supported. if it's add(half,half) then we dont' have a kernel to do that.

Reply:
We don't have kernels to add half tensors? Even on CUDA?

Reply:
added for half cuda.